src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -290,12 +415,33 @@ void KafkaSink::onFinish()
     if (auto err = rd_kafka_flush(producer.get(), 15000 /* time_ms */); err)
         LOG_ERROR(log, "Failed to flush kafka producer, error={}", rd_kafka_err2str(err));
 
-    if (auto err = wb->lastSeenError(); err != RD_KAFKA_RESP_ERR_NO_ERROR)
+    if (auto err = lastSeenError(); err != RD_KAFKA_RESP_ERR_NO_ERROR)
         LOG_ERROR(log, "Failed to send messages, last_seen_error={}", rd_kafka_err2str(err));
 
     /// if flush does not return an error, the delivery report queue should be empty
-    if (!wb->hasNoOutstandings())
-        LOG_ERROR(log, "Not all messsages are sent successfully, expected={} actual={}", wb->outstandings(), wb->acked());
+    if (!hasNoOutstandings())
+        LOG_ERROR(log, "Not all messsages are sent successfully, expected={} actual={}", outstandings(), acked());
+}
+
+void KafkaSink::onMessageDelivery(rd_kafka_t * /* producer */, const rd_kafka_message_t * msg, void * opaque)
+{
+    static_cast<KafkaSink *>(opaque)->onMessageDelivery(msg);
+}
+
+void KafkaSink::onMessageDelivery(const rd_kafka_message_t * msg)
+{
+    /// Can't rely on librdkafka freeing up the payloads with RD_KAFKA_MSG_F_FREE, it will throw:
+    ///   AddressSanitizer: alloc-dealloc-mismatch (operator new [] vs free)
+    char * payload = static_cast<char *>(msg->payload);
+    delete[] payload;
```

Comment:
so, we don't need delete by ourselves when move the memory to librdkafka.

Reply:
Please check exiting code in StreamShard how we manage it

Reply:
Updated, please check.